Add examples to exportparts (#28649)
mdn/content: files/en-us/web/html/global_attributes/exportparts/index.md

Attribute `exportparts` must be placed on a _shadow Host_, which is the element to which the _shadow tree_ is attached. The value of the attribute should be a comma-separated list of part names present in the shadow tree and which should be made available via a DOM outside of the current structure.

## Examples

First let's create a card component that we will then wrap with another component.

```html
<template id="card-component-template">
  <style>
    .base {
      display: grid;
    }
  </style>
  <div class="base" part="base">
    <div part="header"><slot name="header"></slot></div>
    <div part="body"><slot></slot></div>

>     <div part="footer"><slot name="footer"></slot></div>

Author: @estelle I did. I was using it as a "default" slot. But could see where not being explicit here could be confusing.

  "card-component",
  class extends HTMLElement {
    constructor() {
      super(); // Always call super first in constructor
      const cardComponent = document.getElementById(
        "card-component-template",
      ).content;
      const shadowRoot = this.attachShadow({
        mode: "open",
      });
      shadowRoot.appendChild(cardComponent.cloneNode(true));
    }
  },
);
```

### Re-exporting parts

With the above `<card-component>`, we could wrap it with another component and export all of the "parts" as is.

```html
<template id="card-wrapper">
  <card-component exportparts="base, header, body, footer"></card-component>
</template>
```

```js
customElements.define(
  "card-wrapper",
  class extends HTMLElement {
    constructor() {
      super(); // Always call super first in constructor
      const cardWrapper = document.getElementById("card-wrapper").content;
      const shadowRoot = this.attachShadow({
        mode: "open",
      });
      shadowRoot.appendChild(cardWrapper.cloneNode(true));
    }
  },
);
```

Now we can target parts on the `<card-component>` from the `<card-wrapper>` like so:

```css
card-wrapper::part(header) {
  font-weight: bold;
}
```

> ### Renaming / Remapping parts

Reviewer: ```suggestion
### Exposing mapped parts
```

> To rename an exportpartm the syntax looks something like this: `exportparts="<original-name>:<new-name>,<original-name>:<new-name>"`

Reviewer: ```suggestion
To rename exported parts, we include a comma-separated list of mapped parts, with each mapped part including the original name and exported name separated by a colon (`:`): 
```

> And here's the prior `CardWrapper` with the remapping syntax:

Reviewer: ```suggestion
#### HTML

We update the prior `<card-wrapper>` custom element with the remapping syntax (omitting `body` from the exported parts list):
```

we don't explain that `CardWrapper` is the same as `<card-wrapper>`, so let's keep it as an element.

```html
<template id="card-wrapper">
  <card-component
    exportparts="
       base:card__base, 
       header:card__header, 

>        body:card__body, 

Reviewer: ```suggestion
```
Let's continue to omit body so we can show what happens when omitted.

       footer:card__footer
     ">
  </card-component>
</template>
```

> 

Reviewer: ```suggestion

#### JavaScript
```

also include the hidden JS and HTML

```js
customElements.define(
  "card-wrapper",
  class extends HTMLElement {
    constructor() {
      super(); // Always call super first in constructor
      const cardWrapper = document.getElementById("card-wrapper").content;
      const shadowRoot = this.attachShadow({
        mode: "open",
      });
      shadowRoot.appendChild(cardWrapper.cloneNode(true));
    }
  },
);
```

> Now we can target parts on the `<card-component>` from the `<card-wrapper>` like so:

Reviewer: ```suggestion
#### CSS

In targetting parts of the `<card-component>` from within the `<card-wrapper>`, we can only style the exported parts via their exposed part names:

```

```css

> card-wrapper::part(card__header) {
>   font-weight: bold;
> }

Reviewer: ```suggestion
/* selects the exported parts name */
card-wrapper::part(card__header) {
  font-weight: bold;
}
/* selects nothing: these part names were not exported */
card-wrapper::part(footer),
::part(body) {
  font-weight: bold;
}
```

Attribute `exportparts` must be placed on a _shadow Host_, which is the element to which the _shadow tree_ is attached. The value of the attribute should be a comma-separated list of part names present in the shadow tree and which should be made available via a DOM outside of the current structure.

## Examples

First let's create a card component that we will then wrap with another component.

```js
class CardComponent extends HTMLElement {
  connectedCallback() {
    this.attachShadow({ mode: "open" });

>     this.shadowRoot.innerHTML = `

Reviewer: Let's use HTML to create HTML. This is a perfect opportunity to demonstrate using `<template>`

Author: I'm not sure I follow.

Something like this?

```html
<template id="card-component-template">
  <style>
    .base { display: grid };
  </style>
       
   <div class="base" part="base">
     <div part="header"><slot name="header"></slot></div>
     <div part="body"><slot></slot></div>
     <div part="footer"><slot name="footer"></slot></div>
    </div>
</template>
```

```js
class CardComponent extends HTMLElement {
  connectedCallback () {
    this.attachShadow({ mode: "open" })
    const firstChild = this.shadowRoot.firstElementChild
    const content = document.querySelector("#card-component-template").content.cloneNode(true)
    
   this.shadowRoot.innerHTML = ""
    
    this.shadowRoot.appendChild(content)
  }
}
```

Reviewer: Yes. Exactly. Especially the HTML part.

For the HTML pages, where this example resides, I thought it's preferable to use the [`define`](https://developer.mozilla.org/en-US/docs/Web/API/Window/customElements/define) method of the Window's [customElements](https://developer.mozilla.org/en-US/docs/Web/API/Window/customElements) as it's [defined in the WHATWG spec](https://html.spec.whatwg.org/multipage/custom-elements.html#dom-window-customelements). 

```js
customElements.define('card-component',
  class extends HTMLElement {
    constructor() {
      super(); // Always call super first in constructor
      const cardComponent = document.getElementById('card-component-template').content;
      const shadowRoot = this.attachShadow({
        mode: 'open'
      });
      shadowRoot.appendChild(cardComponent.cloneNode(true));
    }
  });
  ```

example pen: https://codepen.io/estelle/pen/YzjNVjM

  <p slot="footer_slot">This is the footer</p>
</card-component>
```

```html
<template id="card-wrapper">
  <card-component exportparts="base, header, body"></card-component>
</template>
```

#### JavaScript

```js hidden
customElements.define(
  "card-component",
  class extends HTMLElement {
    constructor() {
      super(); // Always call super first in constructor
      const cardComponent = document.getElementById(
        "card-component-template",
      ).content;
      const shadowRoot = this.attachShadow({
        mode: "open",
      });
      shadowRoot.appendChild(cardComponent.cloneNode(true));
    }
  },
);
```

```js
customElements.define(
  "card-wrapper",
  class extends HTMLElement {
    constructor() {
      super(); // Always call super first in constructor
      const cardWrapper = document.getElementById("card-wrapper").content;
      const shadowRoot = this.attachShadow({
        mode: "open",
      });
      shadowRoot.appendChild(cardWrapper.cloneNode(true));
    }
  },
);
```

#### CSS

Now we can target parts on the `<card-component>` from the `<card-wrapper>` like so:

```css
::part(body) {
  color: red;
  font-style: italic;
}

> ::part(header, footer) {

Reviewer: ```suggestion
::part(header),
::part(footer) {
```

</template>
```

## Examples

### Basic component

To demonstrate how `exportparts` is used to enable targeting parts within nested components, we will create a component, and then nest it within another component.

#### HTML

First, let's create a card component that we will then wrap with another component. We also use the new element we created, populating the slots with plain text as content.

```html
<template id="card-component-template">
  <style>
    p {
      margin: 0.5em 0;

>     }

Reviewer 2: ```suggestion
    }

    .base {
      width: fit-content;
      border: 2px dashed blue;
    }
```

```js
customElements.define(
  "card-component",
  class extends HTMLElement {
    constructor() {
      super(); // Always call super first in constructor
      const cardComponent = document.getElementById(
        "card-component-template",
      ).content;
      const shadowRoot = this.attachShadow({
        mode: "open",
      });
      shadowRoot.appendChild(cardComponent.cloneNode(true));
    }
  },
);
```

#### CSS

We style parts of the `<card-component>` shadow tree using the {{cssxref("::part")}} pseudo-element:

```css
::part(body) {
  color: red;
  font-style: italic;
}
```

#### Results

{{ EmbedLiveSample('Basic_component', '100%', '160') }}

### Nested component

Continuing the above `<card-component>` example, we create a nested component by wrapping the `<card-component>` within another component; in this case, the `<card-wrapper>` component. We then export the parts from the nested component that we want to make styleable from outside the component's shadow tree with the `exportparts` attribute.

#### HTML

```html hidden
<template id="card-component-template">
  <style>
    p,
    h2 {
      margin: 0.5em 0;

>     }

Reviewer 2: ```suggestion
    }

    .base {
      width: fit-content;
      border: 2px dashed blue;
    }
```

Reviewer: i added this to the component from the main DOM and added :host display block here instead; showing a few more examples of how to bypass the DOM / shadowDOM barrier

  </style>
  <div class="base" part="base">
    <div part="header"><slot name="header_slot"></slot></div>
    <div part="body"><slot name="body_slot"></slot></div>
    <div part="footer"><slot name="footer_slot"></slot></div>
  </div>
</template>
```

```html
<template id="card-wrapper">

>   <card-component exportparts="base, header, body">
>     <span slot="header_slot"><slot name="H"></slot></span>
>     <span slot="body_slot"><slot name="B"></slot></span>
>     <span slot="footer_slot"><slot name="F"></slot></span>
>   </card-component>

Reviewer 2: ```suggestion
  <style>
    .wrapper {
      padding: 1rem;
      border: 2px dashed red;
    }
  </style>
  <div class="wrapper">
    <card-component exportparts="base, header, body">
      <span slot="header_slot"><slot name="H"></slot></span>
      <span slot="body_slot"><slot name="B"></slot></span>
      <span slot="footer_slot"><slot name="F"></slot></span>
    </card-component>
  </div>
```

  <p slot="body_slot">This is the body</p>
  <p slot="footer_slot">This is the footer</p>
</card-component>
```

#### JavaScript

```js hidden
customElements.define(
  "card-component",
  class extends HTMLElement {
    constructor() {
      super(); // Always call super first in constructor
      const cardComponent = document.getElementById(
        "card-component-template",
      ).content;
      const shadowRoot = this.attachShadow({
        mode: "open",
      });
      shadowRoot.appendChild(cardComponent.cloneNode(true));
    }
  },
);
```

```js
customElements.define(
  "card-wrapper",
  class extends HTMLElement {
    constructor() {
      super(); // Always call super first in constructor
      const cardWrapper = document.getElementById("card-wrapper").content;
      const shadowRoot = this.attachShadow({
        mode: "open",
      });
      shadowRoot.appendChild(cardWrapper.cloneNode(true));
    }
  },
);
```

#### CSS

Now we can target parts of the `<card-component>` directly and when nested within a `<card-wrapper>` like so:

```css
::part(body) {
  color: red;
  font-style: italic;
}

::part(header),
::part(footer) {
  font-weight: bold;
}
```

#### Results

> {{ EmbedLiveSample('Nested_component', '100%', '160') }}

Reviewer: ```suggestion
{{ EmbedLiveSample('Nested_component', '100%', '400') }}
```

Note `footer` is not bold when nested, as we did not include it in `exportparts`.

### Exposing mapped parts

To rename exported parts, we include a comma-separated list of mapped parts, with each mapped part including the original name and exported name separated by a colon (`:`):

#### HTML

We update the prior `<card-wrapper>` custom element with the remapping syntax (omitting `body` from the exported parts list):

```html hidden
<template id="card-component-template">
  <style>
    p {
      margin: 0.5em 0;

>     }

Reviewer 2: ```suggestion
    }

    .base {
      width: fit-content;
      border: 2px dashed blue;
    }
```
